jax_distribution/systems/ff_ippo_rware.py

```diff
@@ -0,0 +1,407 @@
+
+    return learner_fn
+
+
+def run_experiment(env: Environment, config: dict) -> dict:
+    """Runs experiment."""
+    # INITIALISE
+    cores_count = len(jax.devices())  # get available TPU cores.
+    num_actions = int(env.action_spec().num_values[0])
+    num_agents = env.action_spec().shape[0]
+    config["NUM_AGENTS"] = num_agents
+    network = ActorCritic(num_actions, config["ACTIVATION"])  # define network.
+    optim = optax.adam(config["LR"])  # define optimiser.
+    rng, rng_e, rng_p = jax.random.split(
+        jax.random.PRNGKey(config["SEED"]), num=3
+    )  # prng keys.
+    total_timesteps = (
+        cores_count
+        * config["NUM_UPDATES"]
+        * config["ROLLOUT_LENGTH"]
+        * config["UPDATE_BATCH_SIZE"]
+        * config["NUM_ENVS"]
+    )
+
+    # INITIALISE NETWORK AND OPTIMISER PARAMS
+    init_x = env.observation_spec().generate_value()
+    # select only obs for a single agent.
+    init_x = jax.tree_util.tree_map(lambda x: x[0], init_x)
+    init_x = jax.tree_util.tree_map(lambda x: x[None, ...], init_x)
+    params = network.init(rng_p, init_x)
+    # initialise optimiser stats.
+    opt_state = optim.init(params)
+
+    vmapped_network_apply_fn = jax.vmap(
+        network.apply, in_axes=(None, 1), out_axes=(1, 1)
+    )
+
+    learn = get_learner_fn(  # get batched iterated update.
+        env, vmapped_network_apply_fn, optim.update, config
+    )
+    learn = jax.pmap(learn, axis_name="device")  # replicate over multiple cores.
+
+    # BROADCAST PARAMS AND OPTIMISER STATE
+    broadcast = lambda x: jnp.broadcast_to(
+        x, (cores_count, config["UPDATE_BATCH_SIZE"]) + x.shape
+    )
+    params = jax.tree_map(broadcast, params)  # broadcast to cores and batch.
+    opt_state = jax.tree_map(broadcast, opt_state)  # broadcast to cores and batch
+
+    # INITIALISE ENVIRONMENT
+    rng, *env_rngs = jax.random.split(
+        rng, cores_count * config["UPDATE_BATCH_SIZE"] * config["NUM_ENVS"] + 1
+    )
+    env_states, timesteps = jax.vmap(env.reset, in_axes=(0))(
+        jnp.stack(env_rngs),
+    )
+    rng, *step_rngs = jax.random.split(
+        rng, cores_count * config["UPDATE_BATCH_SIZE"] + 1
+    )
+
+    # RESHAPE ENVIRONMENT STATES AND TIMESTEPS
+    reshape_step_rngs = lambda x: x.reshape(
+        (cores_count, config["UPDATE_BATCH_SIZE"]) + x.shape[1:]
+    )
+    step_rngs = reshape_step_rngs(jnp.stack(step_rngs))  # add dimension to pmap over.
+
+    reshape_states = lambda x: x.reshape(
+        (cores_count, config["UPDATE_BATCH_SIZE"], config["NUM_ENVS"]) + x.shape[1:]
+    )
+    env_states = jax.tree_util.tree_map(
+        reshape_states, env_states
+    )  # add dimension to pmap over.
+    timesteps = jax.tree_util.tree_map(
+        reshape_states, timesteps
+    )  # add dimension to pmap over.
+
+    # RUN EXPERIMENT
+    with TimeIt(tag="COMPILATION"):
+        learn(params, opt_state, step_rngs, env_states, timesteps)
```

Comment:
This seems like you're running a full experiment before running another, we should definitely tell it to do only a single step of learning and executing, that will compile everything and take much less time.

But in general do we want this here? I understand this is useful when timing stuff, but when doing experiments I don't want to do extra work, the compilation is just part of the runtime, so this probably shouldn't be in the final product.

Reply:
Interesting, yes you made a good point, at the start we did this for testing but I guess we should either make it for one step or remove it completely

Reply:
Sounds good :+1: